File: App/GameClient_HTML5/MorpionRules.js
/** @author William J.D. **/

/*
Naughts and Crosses HTML 5 game example
Copyright (C) 2012 William James Dyce

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

function MorpionRules()
{
	/** ATTRIBUTES **/
	var obj = new Rules();
	var typ = MorpionRules;
	
	/** SUBROUTINES **/
	var hasWon = function(turn, board)
	{
		// count the number which are my colour
		var mine = (turn == Game.WHITE) 
							? Board.CELL_WHITE
							: Board.CELL_BLACK;
		var count = 0;
		
		// check rows
		for(row = 0; row < MorpionBoard.N_ROWS; row++)
		{
			count = 0;
			for(col = 0; col < MorpionBoard.N_COLS; col++)
			{
				if(board.getCell(row, col) != mine)
					break;
				else
					count++;
			}
			if(count == MorpionBoard.N_COLS)
				return true;
		}
		
		// check columns
		for(col = 0; col < MorpionBoard.N_COLS; col++)
		{
			count = 0;
			for(row = 0; row < MorpionBoard.N_ROWS; row++)
			{
				if(board.getCell(row, col) != mine)
					break;
				else
					count++;
			}
			if(count == MorpionBoard.N_ROWS)
				return true;
		}
		
		// check diagonal 1 (downward left to right)
		count = 0;
		for(i = 0; i < MorpionBoard.N_ROWS; i++)
		{
			if(board.getCell(i, i) != mine)
				break;
			else
				count++;
		}
		if(count == MorpionBoard.N_ROWS)
			return true;
			
		// check diagonal 2 (upwards left to right)
		count = 0;
		for(i = 0; i < MorpionBoard.N_COLS; i++)
		{
			if(board.getCell(i, MorpionBoard.N_ROWS-1-i) != mine)
				break;
			else
				count++;
		}
		if(count == MorpionBoard.N_COLS)
			return true;
	}
	
	var isTie = function(board)
	{
		// it's a draw if none of the nine cells are empty
		for(row = 0; row < MorpionBoard.N_ROWS; row++)
			for(col = 0; col < MorpionBoard.N_COLS; col++)
				if(board.getCell(row, col) == Board.CELL_EMPTY)
					return false;
		// no cells are empty
		return true;
	}
	
	/** METHODS **/
	
	obj.firstTurn = function()
	{
		return Game.WHITE;
	}
	
	obj.reset = function(board)
	{
		for(row = 0; row < MorpionBoard.N_ROWS; row++)
			for(col = 0; col < MorpionBoard.N_COLS; col++)
				board.setCell(row, col, Board.CELL_EMPTY);
	}

	obj.isLegalMove = function(row, col, board, turn)
	{
		return (board.getCell(row, col) == Board.CELL_EMPTY);
	}

	obj.performMove = function(row, col, board, turn)
	{
		// set the cell to the new colour
		board.setCell(row, col, (turn == Game.WHITE) 
								? Board.CELL_WHITE 
							  	: Board.CELL_BLACK);
											  
		// report if success has occured
		if(hasWon(turn, board))
			return (turn == Game.WHITE) 
					? Game.VICTORY_WHITE 
				  	: Game.VICTORY_BLACK;
		// report if draw has occured
		else if(isTie(board))
			return Game.DRAW;
		// report next player to move (cycle between the two)
		else
			return (turn == Game.BLACK) 
					? Game.WHITE 
					: Game.BLACK;
	}
	
	/** RETURN INSTANCE **/
	return obj;
}

File: App/GameService/src/java/game/morpion/MorpionRules.java
/*****************
 * @author william
 * @date 21-Mar-2012
 *****************/


package game.morpion;

import game.Board;
import game.Board.Position;
import game.Game;
import game.Game.Player;
import game.Game.State;
import game.Rules;


class MorpionRules extends Rules
{
    /** SINGLETON **/
    
    private static MorpionRules instance = null;
    public static MorpionRules getInstance()
    {
        if(instance == null)
            instance = new MorpionRules();
        return instance;
    }
    private MorpionRules() { }
    
    
    
    /** IMPLEMENT INTERFACE **/
    
    // query
    @Override
    public Player getFirstPlayer() 
    {
        return Player.WHITE;
    }
    
    @Override
    public boolean isDraw(Board board) 
    {
        // return true if every cell is occupied
        return (board.get_n_pieces() == board.get_n_rows()*board.get_n_cols());
    }

    @Override
    public boolean hasWon(Board board, Player player) 
    {
        // local variables
        Position p = new Position(0, 0);
        int count;

        // check vertical '|' alignments
        for(p.col = 0; p.col < board.get_n_cols(); p.col++)
        {
            count = 0;
            for(p.row = 0; p.row < board.get_n_rows(); p.row++)
            {
                if(board.getCellOwner(p) != player)
                    break;
                else
                    count++;
            }
            if(count == board.get_n_rows())
                return true;
        }

        // check horizontal '--' alignments
        for(p.row = 0; p.row < board.get_n_rows(); p.row++)
        {
            count = 0;
            for(p.col = 0; p.col < board.get_n_rows(); p.col++)
            {
                if(board.getCellOwner(p) != player)
                    break;
                else
                    count++;
            }
            if(count == board.get_n_cols())
                return true;
        }
        

        // check for diagonal '\' alignments
        count = 0;
        for(int i = 0; i < board.get_n_rows(); i++)
        {
            p.row = p.col = i;
            if(board.getCellOwner(p) != player)
                break;
            else
                count++;
        }
        if(count == board.get_n_rows())
            return true;

        // check for diagonal '/' alignments
        count = 0;
        for(int i = 0; i < board.get_n_cols(); i++)
        {
            p.row = i;
            p.col = board.get_n_rows()-1-1;
            if(board.getCellOwner(p) != player)
                break;
            else
                count++;
        }
        if(count == board.get_n_cols())
            return true;
        
        // all possibilities have been exhausted
        return false;
    }

    @Override
    public boolean isLegalMove(Position p, Board board, Player player) 
    {
        return (board.getCell(p) == Board.Cell.EMPTY);
    }

    // modification
    @Override
    public Game.State performMove(Position p, Board board, Player player) 
    {
        // set the cell to the new colour
        board.setCellOwner(p, player);

        // check if success has occured
        if(hasWon(board, player))
            return (player == Player.WHITE) ? 
                    State.VICTORY_WHITE : State.VICTORY_BLACK;
        
        // check if draw has occured
        else if(isDraw(board))
                return State.DRAW;
        
        // otherwise report next player to move (cycle between the two)
        else
            return (player == Player.WHITE) ? 
                    State.TURN_BLACK : State.TURN_WHITE;
    }
    
    @Override
    public void reset(Board board) 
    {
        // simply clear the board, that's all there is to it ;)
        board.clear();
    }

}
